extensions/big-data-cluster/src/bigDataCluster/tree/controllerTreeNode.ts

```diff
@@ -0,0 +1,376 @@
+import { IconPath, BdcItemType } from '../constants';
+import { ClusterController, IEndPoint, IControllerError } from '../controller/clusterController';
+
+const localize = nls.loadMessageBundle();
+
+export type ControllerTreeNodeArg = {
+	label: string;
+	treeChangeHandler: IControllerTreeChangeHandler;
+	parent?: ControllerTreeNode;
+	description?: string;
+	nodeType?: string;
+	iconPath?: { dark: string, light: string };
+};
+
+export abstract class ControllerTreeNode extends TreeNode {
+	private _description: string;
+	private _nodeType: string;
+	private _iconPath: { dark: string, light: string };
+	private _treeChangeHandler: IControllerTreeChangeHandler;
+
+	constructor(arg?: ControllerTreeNodeArg) {
+		super(arg);
+		this._treeChangeHandler = arg.treeChangeHandler;
+		this._description = arg.description || arg.label;
+		this._nodeType = arg.nodeType;
+		this._iconPath = arg.iconPath;
+	}
+
+	public async getChildren(): Promise<ControllerTreeNode[]> {
+		return this.children as ControllerTreeNode[];
+	}
+
+	public refresh(): void {
+		super.refresh();
+		this.treeChangeHandler.notifyNodeChanged(this);
+	}
+
+	public getTreeItem(): vscode.TreeItem {
+		let item: vscode.TreeItem = {};
+		item.id = this.id;
+		item.label = this.label;
+		item.collapsibleState = this.isLeaf ?
+			vscode.TreeItemCollapsibleState.None :
+			vscode.TreeItemCollapsibleState.Collapsed;
+		item.iconPath = this._iconPath;
+		item.contextValue = this._nodeType;
+		item.tooltip = this._description;
+		item.iconPath = this._iconPath;
+		return item;
+	}
+
+	public getNodeInfo(): azdata.NodeInfo {
+		return {
+			label: this.label,
+			isLeaf: this.isLeaf,
+			errorMessage: undefined,
+			metadata: undefined,
+			nodePath: this.nodePath,
+			nodeStatus: undefined,
+			nodeType: this._nodeType,
+			iconType: this._nodeType,
+			nodeSubType: undefined
+		};
+	}
+
+	public get description(): string {
+		return this._description;
+	}
+
+	public set description(description: string) {
+		this._description = description;
+	}
+
+	public get nodeType(): string {
+		return this._nodeType;
+	}
+
+	public set nodeType(nodeType: string) {
+		this._nodeType = nodeType;
+	}
+
+	public set iconPath(iconPath: { dark: string, light: string }) {
+		this._iconPath = iconPath;
+	}
+
+	public get iconPath(): { dark: string, light: string } {
+		return this._iconPath;
+	}
+
+	public set treeChangeHandler(treeChangeHandler: IControllerTreeChangeHandler) {
+		this._treeChangeHandler = treeChangeHandler;
+	}
+
+	public get treeChangeHandler(): IControllerTreeChangeHandler {
+		return this._treeChangeHandler;
+	}
+}
+
+export type ControllerRootNodeArg = {
+	treeChangeHandler: IControllerTreeChangeHandler;
+};
+
+export class ControllerRootNode extends ControllerTreeNode {
+
+	constructor(arg?: ControllerRootNodeArg) {
+		super(Object.assign({
+			label: 'root',
+			nodeType: BdcItemType.controllerRoot,
+		}, arg));
+	}
+
+	public async getChildren(): Promise<ControllerNode[]> {
```

Comment:
This isn't async so can we not make it async? Instead just return `Promise.resolve(this.children as ControllerNode[]);`

Reply:
synced offline.